workers/unity/Packages/com.improbable.gdk.core/Worker/CommandMetaData.cs

```diff
@@ -4,31 +4,42 @@
 
 namespace Improbable.Gdk.Core
 {
-    public struct CommandContext<T>
+    public readonly struct CommandContext<T>
     {
-        public Entity SendingEntity;
-        public T Request;
-        public object Context;
+        public readonly Entity SendingEntity;
+        public readonly T Request;
+        public readonly object Context;
+        public readonly long RequestId;
 
-        public CommandContext(Entity sendingEntity, T request, object context)
+        public CommandContext(Entity sendingEntity, T request, object context, long requestId)
         {
             SendingEntity = sendingEntity;
             Request = request;
             Context = context;
+            RequestId = requestId;
         }
     }
 
     public class CommandMetaData
     {
+        // Cache the types needed to instantiate a new CommandMetaData
+        private static List<Type> storageTypes;
+
+        private readonly HashSet<uint> internalRequestIds = new HashSet<uint>();
+
         private readonly Dictionary<uint, Dictionary<uint, ICommandMetaDataStorage>> componentIdToCommandIdToStorage =
             new Dictionary<uint, Dictionary<uint, ICommandMetaDataStorage>>();
 
         private List<CommandIds> requestsToRemove = new List<CommandIds>();
 
         public CommandMetaData()
         {
-            var types = ReflectionUtility.GetNonAbstractTypes(typeof(ICommandMetaDataStorage));
-            foreach (var type in types)
+            if (storageTypes == null)
+            {
+                storageTypes = ReflectionUtility.GetNonAbstractTypes(typeof(ICommandMetaDataStorage));
```

Comment:
Could use the `??` operator to assign, but probably not worth

Reply:
I mildly prefer the explicit check as we normally don't want to perform an assignment, but I'm not too precious about this.